### PR TITLE
TextField - allow to customize cursor color in error state

### LINKED ISSUE #135580: TextFormField - when validation fails, cursor color is assigned to be the same as the error text color
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Steps to reproduce

Hi, I'm reopening this issue: https://github.com/flutter/flutter/issues/134451 because it was locked without solving

1. Run the code sample
2. Tap on the text field to set the focus to it
3. The cursor color is green as expected
4. Write something, no matter what, so the "Invalid" error text will appear in red under the text field

### Expected results

The cursor color should stay green.

### Actual results

The cursor color turns red like the error text color.
If this is intended, then the documentation of TextFormField should be updated, although I don't see the point in keeping this behavior.

### Code sample

<details><summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MainApp());
}

class MainApp extends StatelessWidget {
  const MainApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        body: Form(
          autovalidateMode: AutovalidateMode.always,
          child: Center(
            child: Padding(
              padding: const EdgeInsets.all(20),
              child: TextFormField(
                cursorColor: Colors.green,
                validator: (value) {
                  if (value?.isNotEmpty == true) return 'Invalid';
                  return null;
                },
              ),
            ),
          ),
        ),
      ),
    );
  }
}
```

</details>

### Screenshots or Video

<details>
<summary>Screenshots / Video demonstration</summary>
</details>

### Logs

<details><summary>Logs</summary>

</details>

### Flutter Doctor output

<details><summary>Doctor output</summary>

```console
[!] Flutter (Channel unknown, 3.10.5, on macOS 13.3 22E252 darwin-arm64, locale en-IL)
    ! Flutter version 3.10.5 on channel unknown at /Users/liran/code/flutter
      Currently on an unknown channel. Run `flutter channel` to switch to an official channel.
      If that doesn't fix the issue, reinstall Flutter by following instructions at https://flutter.dev/docs/get-started/install.
    ! Unknown upstream repository.
      Reinstall Flutter by following instructions at https://flutter.dev/docs/get-started/install.
    • Framework revision 796c8ef792 (3 months ago), 2023-06-13 15:51:02 -0700
    • Engine revision 45f6e00911
    • Dart version 3.0.5
    • DevTools version 2.23.1
    • If those were intentional, you can disregard the above warnings; however it is recommended to use "git" directly to perform update checks and upgrades.

[✓] Android toolchain - develop for Android devices (Android SDK version 32.1.0-rc1)
    • Android SDK at /Users/liran/Library/Android/sdk
    • Platform android-34, build-tools 32.1.0-rc1
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.11+0-b60-7772763)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.3.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14E300c
    • CocoaPods version 1.12.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.11+0-b60-7772763)

[✓] VS Code (version 1.81.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.72.0

[✓] Connected device (4 available)
    • sdk gphone64 arm64 (mobile) • emulator-5554             • android-arm64  • Android 14 (API 34) (emulator)
    • Liran's iPhone (mobile)     • 00008101-000265A20C84001E • ios            • iOS 16.6 20G75
    • macOS (desktop)             • macos                     • darwin-arm64   • macOS 13.3 22E252 darwin-arm64
    • Chrome (web)                • chrome                    • web-javascript • Google Chrome 116.0.5845.179

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 1 category.
```

</details>

### PR DESCRIPTION
**Pull request by liranzairi**
The color of the TextField's cursor in error state is the same as the error text color by default. However we should be allowed to customize it

Fixes #135580 

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by Renzo-Olivares** [comment]
Hi @lirantzairi, thank you for the contribution!

I think a better place to put `cursorErrorColor` might be in `InputDecoration` and `InputDecorationTheme` since this is where `errorStyle` lives and this allows someone to define top level themes in `MaterialApp` for all their `TextField`s. What do you think? I do know `cursorColor` belongs to `TextField` and `DefaultSelectionStyle` but i'm leaning towards including this new property on `InputDecoration`.

/cc @LongCatIsLooong for any thoughts.

**Comment by lirantzairi** [comment]
@Renzo-Olivares I was thinking about it, I chose to put it in `TextField` because the concept of a cursor belongs to TextFields and not to other widgets that use `InputDecoration`.  For example a developer that adds a `DropdownButton` with `InputDecoration` might be confused if he sees the option to set a cursor color. It's the same reason why the icon of the dropdown button is managed in the dropdown class and not in the decoration.

Moreover, I would prefer not to connect between `cursorErrorColor` and `errorStyle` because I believe that connection was wrong in the first place due to misunderstanding of Material design guidelines.

Lastly, I feel (and this is a wild guess) that this variable is likely to be used by people who set `cursorColor`, so putting them next to each other is more intuitive.

**Comment by Renzo-Olivares** [comment]
@lirantzairi I see your point! Thanks for the clear explanation. This PR looks good, my main argument for putting it in `InputDecoration` or `DefaultSelectionStyle` is that it would allow the user to override the `cursorErrorColor` for an entire widget subtree rather than a single `TextField`. I'm not completely sold on placing it in either (though leaning towards `DefaultSelectionStyle`), so I think we can hold off on that for now.

File: packages/flutter/test/material/text_form_field_test.dart [hunk 1]
```diff
@@ -1481,27 +1481,59 @@ void main() {
...
+    await runTest(
+      buildWidget(errorStyleColor: errorStyleColor),
+      expectedColor: errorStyleColor,
+    );
+    await runTest(
+      buildWidget(cursorErrorColor: cursorErrorColor),
+      expectedColor: cursorErrorColor,
+    );
+    await runTest(
+      buildWidget(errorStyleColor: errorStyleColor, cursorErrorColor: cursorErrorColor),
```
Comment (Renzo-Olivares) [inline packages/flutter/test/material/text_form_field_test.dart:1534]: nit: formatting, put each parameter on their own line and the last parameter should have a trailing comma.

**Review by Renzo-Olivares — APPROVED** [review]
Had one small comment but this looks good to me. This will need a secondary reviewer before it is merged, and you need to sync your PR up to `master` to fix the failing tests.

**Comment by lirantzairi** [comment]
@Renzo-Olivares I fixed the code format according to your comment and rebased on `master` so that all tests pass now. Thank you for the review :)

**Review by bleroux — APPROVED** [review]
LGTM!

**Comment by tenhobi** [comment]
Hi, can I ask you when this PR will be in stable Flutter? Or in general, how can I tell? Thanks a lot!

**Comment by MilanVilov** [comment]
Thank you for this PR, I lost almost an hour looking into this issue until I figured it out. Awesome work! 🦾

**Comment by lirantzairi** [comment]
Thanks @MilanVilov :)

**Comment by gloinho** [comment]
this works on flutter 3.16.6?

**Comment by bleroux** [comment]
> this works on flutter 3.16.6?

Not on 3.16, this fix was included in 3.19.0.

**Pull request merged**